starlight/utils/search.py

```diff
@@ -0,0 +1,351 @@
+from __future__ import annotations
+
+import difflib
+
+from operator import attrgetter
+from typing import (
+    overload,
+    TYPE_CHECKING,
+    TypeVar,
+    Generic,
+)
+
+T = TypeVar('T')
+
+if TYPE_CHECKING:
+    from typing import (
+        List,
+        Dict,
+        Tuple,
+        Any,
+        AsyncIterable,
+        Callable,
+        Coroutine,
+        Iterable,
+        Union,
+    )
+
+    Coro = Coroutine[Any, Any, T]
+    _Iter = Union[Iterable[T], AsyncIterable[T]]
+
+__all__ = (
+    'SearchFilter',
+    'ContainsFilter',
+    'FuzzyFilter',
+    'Contains',
+    'Fuzzy',
+    'search',
+)
+
+
+class SearchFilter(Generic[T]):
+    """Base class for filtering with :func:`search`.
+
+    Parameters
+    ------------
+    query: Any
+        The query value to filter by.
+    """
+    def __init__(self, query: T):
+        self.query: T = query
+
+    def filter(self, value: Any, /) -> Any:
+        """Filters the value from the iterable. The return value will be included
+        if it is truthy or excluded if it is falsy. The return value is used to sort
+        the item.
+
+        Subclasses should override this method.
+
+        Parameters
+        ------------
+        value: Any
+            The value of the attribute from the item in iterable.
+
+        Returns
+        --------
+        Any
+            The value to sort this item by.
+
+        """
+        raise NotImplementedError
+
+
+class ContainsFilter(SearchFilter[Any]):
+    """Basic filter that checks if ``query`` is in the attribute value.
+
```

Comment:
Maybe show an equivalent 
```suggestion
    .. code-block:: python3
        starlight.search(items, my_attr=ContainsFilter('my_value'))
    Equivalent of 
    .. code-block:: python3
        [item for item in filter(lambda item: 'my_value' in item.my_attr, items)]
```

Reply:
I think showing an equivalent should give an idea to the user, because the syntax could make it unfamiliar.